app/features/receive/cashu-token-swap-hooks.ts

```diff
@@ -0,0 +1,309 @@
+  });
+
+  useOnCashuTokenSwapChange({
+    onCreated: (swap) => {
+      pendingSwapsCache.add(swap);
+    },
+    onUpdated: (swap) => {
+      tokenSwapCache.updateIfExists(swap);
+
+      const isSwapStillPending = swap.state === 'PENDING';
+      if (isSwapStillPending) {
+        pendingSwapsCache.update(swap);
+      } else {
+        pendingSwapsCache.remove(swap);
+      }
+    },
+  });
+
+  return data;
+}
+
+function useFinalizeTokenSwap(swaps: CashuTokenSwap[]) {
+  const tokenSwapService = useCashuTokenSwapService();
+  const accountsCache = useAccountsCache();
+
+  useQueries({
+    queries: swaps
+      .filter((swap) => swap.state === 'PENDING')
+      .map((swap) => ({
+        queryKey: ['finalize-cashu-token-swap', swap.id],
+        queryFn: async () => {
+          try {
+            const account = accountsCache.get(swap.accountId);
+            if (!account || account.type !== 'cashu') {
+              throw new Error(`Account not found for id: ${swap.accountId}`);
+            }
+            console.log('Finalizing token swap', swap);
+            return tokenSwapService.finalizeSwap(account, swap);
+          } catch (error) {
+            console.error('Error finalizing token swap', error);
+            throw error;
+          }
+        },
+        refetchInterval: 10000,
+        refetchIntervalInBackground: true,
+      })),
+  });
+}
+
+function useOnCashuTokenSwapChange({
+  onCreated,
+  onUpdated,
+}: {
+  onCreated: (swap: CashuTokenSwap) => void;
+  onUpdated: (swap: CashuTokenSwap) => void;
+}) {
+  const cashuCryptography = useCashuCryptography();
+  const onCreatedRef = useLatest(onCreated);
+  const onUpdatedRef = useLatest(onUpdated);
+
+  useEffect(() => {
+    const channel = boardwalkDb
+      .channel('cashu-token-swaps')
+      .on(
+        'postgres_changes',
+        {
+          event: '*',
+          schema: 'wallet',
+          table: 'cashu_token_swaps',
+        },
+        async (
+          payload: RealtimePostgresChangesPayload<BoardwalkDbCashuTokenSwap>,
+        ) => {
+          if (payload.eventType === 'INSERT') {
+            console.log('Token swap created', payload.new);
+            const swap = await CashuTokenSwapRepository.toTokenSwap(
+              payload.new,
+              cashuCryptography.decrypt,
+            );
+            onCreatedRef.current(swap);
+          } else if (payload.eventType === 'UPDATE') {
+            console.log('Token swap updated', payload.new);
+            const swap = await CashuTokenSwapRepository.toTokenSwap(
+              payload.new,
+              cashuCryptography.decrypt,
+            );
+            onUpdatedRef.current(swap);
+          }
+        },
+      )
+      .subscribe();
+
+    return () => {
+      channel.unsubscribe();
+    };
+  }, [cashuCryptography]);
+}
+
+export function useTrackCashuTokenSwaps() {
+  const pendingSwaps = usePendingCashuTokenSwaps();
+  useFinalizeTokenSwap(pendingSwaps);
```

Comment:
I am not sure we need to do it like this here. The thing is that here the process is:
1. create a swap in our db in pending state
2. make actual swap
3. store proofs in our db and update a swap to completed state

Every step happens immediately after the previous one unlike with those quotes where complete step needs to wait for a signal from mint that payment was done.

Thus I am thinking if we should be doing all these steps from one place at the time of swap creation and then only have "background process" handle ones that are stuck in pending state on app start